Lib/_pyio.py

```diff
@@ -1654,14 +1655,18 @@ def readall(self):
         """
         self._checkClosed()
         self._checkReadable()
-        bufsize = DEFAULT_BUFFER_SIZE
-        try:
-            pos = os.lseek(self._fd, 0, SEEK_CUR)
-            end = os.fstat(self._fd).st_size
-            if end >= pos:
-                bufsize = end - pos + 1
-        except OSError:
-            pass
+        if self._estimated_size <= 0:
+            bufsize = DEFAULT_BUFFER_SIZE
+        else:
+            bufsize = self._estimated_size + 1
```

Comment:
What is the purpose of the "+1"? It may overallocate 1 byte which is inefficient.

Reply:
The read loop currently needs to do a `os.read()`  / `_py_Read` which is a single byte which returns 0 size to find the end of the file and exit the loop. The very beginning of that loop does a check for "if buffer is full, grow buffer" so not over-allocating by one byte results in a much bigger allocation by that. In the `_io` case it then shrinks it back down at the end, whereas in the `_pyio` case the EOF read is never appended.

Could avoid the extra byte by writing a specialized "read known size" (w/ fallback to "read until EOF"), but was trying to avoid making more variants of the read loop and limit risk a bit.

As an aside: the `_pyio` implementation seems to have a lot of extra memory allocation and copy in the default case because `os.read()` internally allocates a buffer which it then copies into its `bytearray`...